encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -166,15 +178,42 @@
     ObjectDecoderContextImpl<T> decoderCtx = getObjectDecodersCtx(classToken);
     reader.beginObject();
     while (reader.hasNext()) {
-      String fieldName = reader.nextName();
-      FieldRef<?> fieldRef = decoderCtx.getFieldRef(fieldName);
-      creationCtx.put(fieldRef, decode(fieldRef.getTypeToken()));
+      decodeField(decoderCtx, creationCtx);
     }
     reader.endObject();
     decoderCtx.decodeInlineObjIfAny(creationCtx);
     return creationCtx;
   }
 
+  private <T> void decodeField(ObjectDecoderContextImpl<T> decoderCtx, CreationContextImpl creationCtx) throws IOException {
+    String fieldName = reader.nextName();
+    FieldRef<?> fieldRef = decoderCtx.getFieldRef(fieldName);
+    FieldDescriptor fieldDescriptor = decoderCtx.getFieldDescriptors(fieldName);
+    Object val = decodeField(fieldRef.getTypeToken(), fieldDescriptor);
+    creationCtx.put(fieldRef, val);
+  }
+
+  private <TField, U extends Annotation> TField decodeField(TypeToken<TField> fieldTypeToken, FieldDescriptor fieldDescriptor) throws IOException {
+    TField val = decode(fieldTypeToken);
+
+    //apply annotation modifier
+    for (Class<? extends Annotation> annotationType: fieldDescriptor.getProperties().keySet()) {
```

Comment:
instead of iterating the types, I suggest you iterate the values, then you don't need to pass the `fieldDescriptor` to `applyFieldModifier()`.

To get the type of the annotation you can just do `annotation.annotationType()`.